packages/front-end/components/Settings/EditDataSource/AutomaticDimension/UpdateAutomaticDimensions.tsx

```diff
@@ -0,0 +1,299 @@
+import React, { FC, Fragment, useCallback, useEffect, useState } from "react";
+import {
+  DataSourceInterfaceWithParams,
+  ExposureQuery,
+} from "back-end/types/datasource";
+import cloneDeep from "lodash/cloneDeep";
+import {
+  AutomaticDimensionInterface,
+  AutomaticDimensionResult,
+} from "back-end/src/types/Integration";
+import { ago, datetime } from "shared/dates";
+import { QueryStatus } from "back-end/types/query";
+import { AUTOMATIC_DIMENSION_OTHER_NAME } from "shared/constants";
+import { useAuth } from "@/services/auth";
+import useApi from "@/hooks/useApi";
+import RunQueriesButton, {
+  getQueryStatus,
+} from "@/components/Queries/RunQueriesButton";
+import ViewAsyncQueriesButton from "@/components/Queries/ViewAsyncQueriesButton";
+import Modal from "@/components/Modal";
+
+const smallPercentFormatter = new Intl.NumberFormat(undefined, {
+  style: "percent",
+  maximumFractionDigits: 0,
+});
+
+export async function setExposureId(
+  exposureQuery: ExposureQuery,
+  dataSource: DataSourceInterfaceWithParams,
+  apiCall: <T>(
+    url: string | null,
+    options?: RequestInit | undefined
+  ) => Promise<T>,
+  setId: (id: string) => void,
+  mutate: () => void
+): Promise<void> {
+  if (exposureQuery.automaticDimensionId) {
+    setId(exposureQuery.automaticDimensionId);
+    await mutate();
+  } else {
+    try {
+      const res = await apiCall<{
+        automaticDimension: AutomaticDimensionInterface;
+      }>(
+        `/automatic-dimension/datasource/${dataSource.id}/${exposureQuery.id}`
+      );
+      if (res?.automaticDimension?.id) {
+        setId(res.automaticDimension.id);
+        await mutate();
+      }
+    } catch (e) {
+      console.error(e);
+    }
+  }
+}
+
+type UpdateAutomaticDimensionModalProps = {
+  exposureQuery: ExposureQuery;
+  dataSource: DataSourceInterfaceWithParams;
+  close: () => void;
+  onSave: (exposureQuery: ExposureQuery) => void;
+};
+
+export const UpdateAutomaticDimensionsModal: FC<UpdateAutomaticDimensionModalProps> = ({
+  exposureQuery,
+  dataSource,
+  close,
+  onSave,
+}) => {
+  const { apiCall } = useAuth();
+  const [id, setId] = useState<string | null>(null);
+  const { data, error, mutate } = useApi<{
+    automaticDimension: AutomaticDimensionInterface;
+  }>(`/automatic-dimension/${id}`);
+
+  useEffect(() => {
+    setExposureId(exposureQuery, dataSource, apiCall, setId, mutate);
```

Comment:
This feels clunky and could lead to unexpected api calls. What are you trying to accomplish, in terms of ux or data flow?